File: ops/cmd/install-daemons.js
/**
 * Install apple-docs LaunchDaemons + sudoers drop-in. Idempotent.
 * Must be invoked as root (the bash predecessor exits 1 on non-root;
 * we mirror that).
 *
 * Flow:
 *   1. Render every *.tpl from ops/.env (delegates to render-all).
 *   2. Unload any stale user-session LaunchAgents.
 *   3. Ensure caddy is installed for the operator user.
 *   4. Bootout the existing system daemons.
 *   5. Optional cleanup of legacy labels from a prior deployment.
 *   6. `install -o root -g wheel -m 644` each plist into
 *      /Library/LaunchDaemons.
 *   7. Bootstrap (or kickstart-fallback) each app daemon.
 *   8. Bring the cloudflared tunnels up.
 *   9. Validate + install the sudoers drop-in via `visudo -cf` then
 *      `install -m 440 …`.
 *  10. Sleep 8s, then run the smoke battery.
 *  11. Optional cleanup of compatibility symlinks pointing at the
 *      ops directory.
 *
 * The "as root" steps shell out via `runCmd` because they need real
 * `sudo`/`install`/`launchctl` privileges. Tests inject `runCmd` +
 * `runCmdAllowFailure` fakes so the suite never tries to actually
 * touch /Library/LaunchDaemons.
 */

import { existsSync } from 'node:fs'
import { join } from 'node:path'
import { loadEnv } from '../lib/env.js'
import { createLogger } from '../lib/logger.js'
import { bootout, bootstrapOrKick, isLoaded, kickstart } from '../lib/launchctl.js'
import { runCmd, runCmdAllowFailure } from '../lib/run-cmd.js'
import runRenderAll from './render-all.js'
import runSmokeTest from './smoke-test.js'

const APP_LABEL_KEYS = ['LABEL_PROXY', 'LABEL_WEB', 'LABEL_MCP', 'LABEL_WATCHDOG']
const ALL_LABEL_KEYS = [...APP_LABEL_KEYS, 'LABEL_TUNNEL_WEB', 'LABEL_TUNNEL_MCP']

/**
 * @param {{ args?: string[], env?: Record<string,string>,
 *           envLoader?: () => any, logger?: any,
 *           deps?: { isRoot?: () => boolean,
 *                    runCmd?: typeof runCmd,
 *                    runCmdAllowFailure?: typeof runCmdAllowFailure,
 *                    bootout?: typeof bootout,
 *                    bootstrapOrKick?: typeof bootstrapOrKick,
 *                    isLoaded?: typeof isLoaded,
 *                    kickstart?: typeof kickstart,
 *                    exists?: (p: string) => boolean,
 *                    renderAll?: typeof runRenderAll,
 *                    smokeTest?: typeof runSmokeTest,
 *                    sleep?: (ms: number) => Promise<void> } }} ctx
 */
export default async function runInstallDaemons(ctx = {}) {
  const logger = ctx.logger ?? createLogger()
  const deps = ctx.deps ?? {}
  const isRoot = deps.isRoot ?? (() => process.getuid?.() === 0)
  if (!isRoot()) {
    logger.error('install-daemons: must be run as root (sudo).')
    return 1
  }

  const env = ctx.envLoader ? ctx.envLoader() : loadEnv()
  const runner = deps.runCmd ?? runCmd
  const runAllow = deps.runCmdAllowFailure ?? runCmdAllowFailure
  const exists = deps.exists ?? existsSync
  const sleep = deps.sleep ?? ((ms) => new Promise(r => setTimeout(r, ms)))
  const renderAll = deps.renderAll ?? runRenderAll
  const smokeTest = deps.smokeTest ?? runSmokeTest

  // 1. Render templates as the operator user. We delegate to
  // render-all.js — when in-process the loaded env is reused.
  logger.say('=== rendering templates ===')
  const rcRender = await renderAll({ args: [], envLoader: () => env, logger })
  if (rcRender !== 0) return rcRender

  // 2. Strip stale user-session LaunchAgents. These would have been
  // installed by an earlier `launchctl bootstrap gui/<uid>` path; the
  // current daemon model is system-domain only.
  logger.say('=== unloading stale user LaunchAgents ===')
  const uidLine = await runAllow(['/usr/bin/id', '-u', env.vars.USER_NAME], { deadlineMs: 5_000 })
  const uid = uidLine.stdout.trim()
  for (const key of ALL_LABEL_KEYS) {
    const label = env.vars[key]
    await runAllow([
      '/usr/bin/sudo', '-u', env.vars.USER_NAME, '/bin/launchctl', 'bootout', `gui/${uid}/${label}`,
    ], { deadlineMs: 5_000 })
    await runAllow([
      '/bin/rm', '-f', `/Users/${env.vars.USER_NAME}/Library/LaunchAgents/${label}.plist`,
    ], { deadlineMs: 5_000 })
  }

  // 3. Caddy presence — delegate to a small sudo'd run-as-user step.
  // (The bash version runs `ensure-caddy.sh` as the operator user;
  // we do the same via sudo -u.)
  logger.say('=== ensuring caddy is installed ===')
  try {
    await runner([
      '/usr/bin/sudo', '-u', env.vars.USER_NAME,
      env.bunBin, join(env.opsDir, '..', 'ops/cli.js'), 'proxy', 'validate',
    ], { deadlineMs: 60_000 })
  } catch (err) {
    // ensure-caddy in the bash flow was a soft pre-check; we warn
    // rather than fail-hard so a missing Caddyfile.tpl render slip
    // doesn't gate the whole install.
    logger.warn(`caddy validation failed: ${err?.message ?? err}`)
  }

  // 4. Bootout existing system daemons.
  logger.say('=== unloading app daemons before reinstall ===')
  const launchctl = {
    bootout: deps.bootout ?? bootout,
    bootstrapOrKick: deps.bootstrapOrKick ?? bootstrapOrKick,
    isLoaded: deps.isLoaded ?? isLoaded,
    kickstart: deps.kickstart ?? kickstart,
  }
  for (const key of APP_LABEL_KEYS) {
    await launchctl.bootout(env.vars[key], { runCmdAllowFailure: runAllow })
  }

  // 5. Legacy cleanup.
  if (env.vars.LEGACY_LAUNCHD_LABELS) {
    logger.say('=== removing legacy launchd labels ===')
    const labels = env.vars.LEGACY_LAUNCHD_LABELS.split(',').map(s => s.trim()).filter(Boolean)
    for (const label of labels) {
      logger.say(`  legacy: ${label}`)
      await launchctl.bootout(label, { runCmdAllowFailure: runAllow })
      await runAllow(['/bin/rm', '-f', `/Library/LaunchDaemons/${label}.plist`], { deadlineMs: 5_000 })
    }
  }

  // 6. Install rendered plists.
  logger.say('=== installing plists to /Library/LaunchDaemons ===')
  for (const key of ALL_LABEL_KEYS) {
    const label = env.vars[key]
    const src = join(env.opsDir, 'launchd', `${label}.plist`)
    if (!exists(src)) {
      logger.warn(`skip: ${src} missing`)
      continue
    }
    await runner([
      '/usr/bin/install', '-o', 'root', '-g', 'wheel', '-m', '644',
      src, '/Library/LaunchDaemons/',
    ], { deadlineMs: 5_000 })
  }

  // 7. Bootstrap app daemons.
  logger.say('=== bootstrapping app daemons ===')
  for (const key of APP_LABEL_KEYS) {
    const label = env.vars[key]
    const plist = `/Library/LaunchDaemons/${label}.plist`
    if (!exists(plist)) continue
    await launchctl.bootstrapOrKick(label, plist, {
      runCmdAllowFailure: runAllow, runCmd: runner, logger,
    })
  }

  // 8. Cloudflared tunnels.
  logger.say('=== ensuring cloudflared daemons are loaded ===')
  for (const key of ['LABEL_TUNNEL_WEB', 'LABEL_TUNNEL_MCP']) {
    const label = env.vars[key]
    const plist = `/Library/LaunchDaemons/${label}.plist`
    if (!exists(plist)) continue
    if (await launchctl.isLoaded(label, { runCmd: runAllow })) {
      await launchctl.kickstart(label, { runCmd: runner })
    } else {
      await launchctl.bootstrapOrKick(label, plist, {
        runCmdAllowFailure: runAllow, runCmd: runner, logger,
      })
    }
  }

  // 9. Sudoers drop-in.
  logger.say('=== validating + installing sudoers drop-in ===')
  const sudoersStem = env.vars.LABEL_PREFIX.replace(/\./g, '_')
  const sudoersFile = `/etc/sudoers.d/${sudoersStem}-launchctl`
  const renderedSudoers = join(env.opsDir, 'launchd', 'sudoers.apple-docs-launchctl')
  await runner(['/usr/sbin/visudo', '-cf', renderedSudoers], { deadlineMs: 5_000 })
  await runner([
    '/usr/bin/install', '-o', 'root', '-g', 'wheel', '-m', '440',
    renderedSudoers, sudoersFile,
  ], { deadlineMs: 5_000 })

  // 10. Smoke.
  logger.say('=== waiting 8s for tunnels and services to settle ===')
  await sleep(8_000)
  logger.say('=== smoke tests ===')
  const rcSmoke = await smokeTest({ envLoader: () => env, logger })
  if (rcSmoke !== 0) logger.warn('one or more smoke tests failed')
  return 0
}

File: swift/Sources/ADOps/InstallDaemons.swift
// `ops install` — render + install launchd plists and the sudoers drop-in.
// Native port of ops/cmd/install-daemons.js. PREPARE-ONLY here: privileged steps
// (install / visudo / launchctl) all go through the injected CommandRunner +
// Launchctl, so `--dry-run` traces them without executing. Must be run as root
// on the live host (the CLI's `--execute`).

private import Foundation

private let appLabelKeys = ["LABEL_PROXY", "LABEL_WEB", "LABEL_MCP", "LABEL_WATCHDOG", "LABEL_AUTOROLL"]
private let allLabelKeys = appLabelKeys + ["LABEL_TUNNEL_WEB", "LABEL_TUNNEL_MCP"]

public enum InstallDaemons {
    /// Injected seams. `renderAll`/`smoke` default to the native ADOps paths.
    public struct Deps: Sendable {
        public var fs: any OpsFileSystem
        public var runner: any CommandRunner
        public var launchctl: Launchctl
        public var http: any HTTPProbing
        public var sleep: @Sendable (Int) async -> Void
        public var isRoot: @Sendable () -> Bool

        public init(
            fs: any OpsFileSystem, runner: any CommandRunner, launchctl: Launchctl,
            http: any HTTPProbing, sleep: @escaping @Sendable (Int) async -> Void = SmokeDeps.systemSleep,
            isRoot: @escaping @Sendable () -> Bool
        ) {
            self.fs = fs
            self.runner = runner
            self.launchctl = launchctl
            self.http = http
            self.sleep = sleep
            self.isRoot = isRoot
        }
    }

    public static func run(env: LoadedEnv, deps: Deps, logger: any OpsLogging) async -> Int32 {
        guard deps.isRoot() else {
            logger.error("install-daemons: must be run as root (sudo).")
            return 1
        }
        let runner = deps.runner
        let launchctl = deps.launchctl

        // 1. Render templates.
        logger.say("=== rendering templates ===")
        let renderOutcome = RenderAll.run(env: env, mode: .write, fs: deps.fs, logger: logger)
        if renderOutcome.exitCode != 0 { return renderOutcome.exitCode }

        // 2. Strip stale user-session LaunchAgents.
        logger.say("=== unloading stale user LaunchAgents ===")
        let userName = env.vars["USER_NAME"] ?? ""
        let uidResult = try? await runner.runAllowFailure(
            ["/usr/bin/id", "-u", userName], options: RunCmdOptions(deadlineMs: 5_000))
        let uid = (uidResult?.stdout ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        for key in allLabelKeys {
            let label = env.vars[key] ?? ""
            _ = try? await runner.runAllowFailure(
                ["/usr/bin/sudo", "-u", userName, "/bin/launchctl", "bootout", "gui/\(uid)/\(label)"],
                options: RunCmdOptions(deadlineMs: 5_000))
            _ = try? await runner.runAllowFailure(
                ["/bin/rm", "-f", "/Users/\(userName)/Library/LaunchAgents/\(label).plist"],
                options: RunCmdOptions(deadlineMs: 5_000))
        }

        // 3. Caddy presence (soft pre-check).
        logger.say("=== ensuring caddy is installed ===")
        do {
            _ = try await runner.run(
                [
                    "/usr/bin/sudo", "-u", userName, env.bunBin,
                    joinPath(env.opsDir, "cli.js"), "proxy", "validate"
                ], options: RunCmdOptions(deadlineMs: 60_000))
        } catch {
            logger.warn("caddy validation failed: \(error)")
        }

        // 4. Bootout existing app daemons.
        logger.say("=== unloading app daemons before reinstall ===")
        for key in appLabelKeys { _ = try? await launchctl.bootout(env.vars[key] ?? "") }

        // 5. Legacy cleanup.
        if let legacy = env.vars["LEGACY_LAUNCHD_LABELS"], !legacy.isEmpty {
            logger.say("=== removing legacy launchd labels ===")
            for label in legacy.split(separator: ",").map({ $0.trimmingCharacters(in: .whitespaces) })
                .filter({ !$0.isEmpty })
            {
                logger.say("  legacy: \(label)")
                _ = try? await launchctl.bootout(label)
                _ = try? await runner.runAllowFailure(
                    ["/bin/rm", "-f", "/Library/LaunchDaemons/\(label).plist"],
                    options: RunCmdOptions(deadlineMs: 5_000))
            }
        }

        // 6. Install rendered plists.
        logger.say("=== installing plists to /Library/LaunchDaemons ===")
        for key in allLabelKeys {
            let label = env.vars[key] ?? ""
            let src = joinPath(joinPath(env.opsDir, "launchd"), "\(label).plist")
            guard deps.fs.exists(src) else {
                logger.warn("skip: \(src) missing")
                continue
            }
            _ = try? await runner.run(
                ["/usr/bin/install", "-o", "root", "-g", "wheel", "-m", "644", src, "/Library/LaunchDaemons/"],
                options: RunCmdOptions(deadlineMs: 5_000))
        }

        // 7. Bootstrap app daemons.
        logger.say("=== bootstrapping app daemons ===")
        for key in appLabelKeys {
            let label = env.vars[key] ?? ""
            let plist = "/Library/LaunchDaemons/\(label).plist"
            guard deps.fs.exists(plist) else { continue }
            _ = try? await launchctl.bootstrapOrKick(label, plistPath: plist)
        }

        // 8. Cloudflared tunnels.
        logger.say("=== ensuring cloudflared daemons are loaded ===")
        for key in ["LABEL_TUNNEL_WEB", "LABEL_TUNNEL_MCP"] {
            let label = env.vars[key] ?? ""
            let plist = "/Library/LaunchDaemons/\(label).plist"
            guard deps.fs.exists(plist) else { continue }
            if (try? await launchctl.isLoaded(label)) == true {
                _ = try? await launchctl.kickstart(label)
            } else {
                _ = try? await launchctl.bootstrapOrKick(label, plistPath: plist)
            }
        }

        // 9. Sudoers drop-in.
        logger.say("=== validating + installing sudoers drop-in ===")
        let stem = (env.vars["LABEL_PREFIX"] ?? "").replacingOccurrences(of: ".", with: "_")
        let renderedSudoers = joinPath(joinPath(env.opsDir, "launchd"), "sudoers.apple-docs-launchctl")
        _ = try? await runner.run(
            ["/usr/sbin/visudo", "-cf", renderedSudoers], options: RunCmdOptions(deadlineMs: 5_000))
        _ = try? await runner.run(
            [
                "/usr/bin/install", "-o", "root", "-g", "wheel", "-m", "440", renderedSudoers,
                "/etc/sudoers.d/\(stem)-launchctl"
            ], options: RunCmdOptions(deadlineMs: 5_000))

        // 10. Smoke.
        logger.say("=== waiting 8s for tunnels and services to settle ===")
        await deps.sleep(8_000)
        logger.say("=== smoke tests ===")
        let smokeCode = await SmokeTest.run(
            env: env, deps: SmokeDeps(http: deps.http, sleep: deps.sleep), logger: logger)
        if smokeCode != 0 { logger.warn("one or more smoke tests failed") }
        return 0
    }
}
